Add support for bus state propagation through socketcan

maxvankessel opened the pull request:
In #19679, the can bus state can be monitored. So this PR is based on #19679 .
This is an update for the socketcan driver, to send error frames to userspace.
And enabling support for filtering on error frames.

jukkar commented on subsys/net/lib/sockets/sockets_can.c:
```diff
@@ -167,9 +168,7 @@ static void zcan_received_cb(struct net_context *ctx, struct net_pkt *pkt,
 		}
 	}
 
-	if (clone && clone != pkt) {
```
Why did you remove the clone variable, it is used if we have multiple socket_can receives?

maxvankessel replied:
I will test this once more, but I was confident that `pkt` needed to be unref'ed otherwise the `net_pkt_` buffer will fill up and won't be cleared.

jukkar replied:
The clone was important when we have multiple socket_can listeners. This can be tested with samples/net/sockets/can sample app and setting CONFIG_NET_SOCKETS_CAN_RECEIVERS=2

maxvankessel replied:
I've tested it again. And I think I know what's wrong.
In my opinion the initial code is invalid.

I'm seeing the problem when only one receiver is configured.
And if it's configured with a filter which does **not**  matches the current packet.
The packet is not unreferenced.

If no receiver is found then `clone == NULL`  and if statement is ignored.
```
if (clone && clone != pkt) {
	net_pkt_unref(pkt);
}
```
So I think the statement should be:
```
if (!clone || clone != pkt) {
	net_pkt_unref(pkt);
}
```
If `clone` isn't set, this means no receiver is found. Then `pkt` has no value for the rest of the application and should not hold a place in the buffer because no one is waiting for this `pkt`
But if it has value, and it's cloned with `net_pkt_clone`. The orignal may be deleted, because it's cloned.

What do you think @jukkar?

jukkar replied:
> I'm seeing the problem when only one receiver is configured.
> And if it's configured with a filter which does **not** matches the current packet.
> The packet is not unreferenced.
> 
> If no receiver is found then `clone == NULL` and if statement is ignored.
> 

There should be at least one receiver in this case as otherwise there would be no socket and this piece of code would not be called in the first place.

The original code tried to check whether there were other listeners for the same socket, and if there were, it cloned the pkt and delivered it to recipient. If there was more that one recipient, we cloned the packet and left the original pkt to be freed at the end of the func. This behaviour could be optimized so we would not need to unref the packet in this func.

maxvankessel replied:
I think you don't understand. Let me start from the beginning.

It's not possible to filter on socketcan error frames in the can hardware driver.
Because it's a socketcan frame and not a can frame, its generated by the (inline) sockcan_generic module. Thats why it must be filtered in the net layer. 

If there are sockets binded to the interface, it's able to receive all frames or (almost) none.
An error frame does always trigger `zcan_received_cb`, if no receiver want's to receive this frame it must be unreferenced. We can't deny existence.

The original code relies on the hardware filter of the can driver. Only packages registered for are distributed to the contexts/sockets. Since the use of error frames, which are eventually mock frames to notify userspace, the `zcan_received_cb` is called from the socketcan_generic module (through `net_recv_data`) even if no package is received from the hardware driver. This frame is matched in `zcan_received_cb` (essentially this functionality was already present), if no one wants this frame it need to be unreferenced.

There were no excess frames in the original driver, because all was filtered by hardware. So all could be dispatched, and there was no direct problem. But if an excess frame is inserted through the socketcan_generic module. You would see the following error eventually:

 ` <err> can_driver: Failed to obtain RX buffer` from the socketcan_generic module.

maxvankessel commented on drivers/can/socket_can_generic.h:
```diff
@@ -39,40 +41,104 @@ static inline void socket_can_iface_init(struct net_if *iface)
 	 * net_if.c:net_if_tx() so that the net_pkt will get released.
 	 */
-	return -ret;
+	return ret;
```
Valid return codes of can_send are:
```
/* CAN_TX_* are the error flags from tx_callback and send.*/
/** send successfully */
#define CAN_TX_OK       (0)
/** general send error */
#define CAN_TX_ERR      (-2)
/** bus arbitration lost during sending */
#define CAN_TX_ARB_LOST (-3)
/** controller is in bus off state */
#define CAN_TX_BUS_OFF  (-4)
/** unexpected error */
#define CAN_TX_UNKNOWN  (-5)
```
So return value is negative, we would not make it negative again.

alexanderwachter replied:
@jukkar @maxvankessel The error codes were changed in the last API change. Apparently we forgot to change it here too.

maxvankessel commented on drivers/can/socket_can_generic.h:
```diff
@@ -83,21 +149,28 @@ static inline int socket_can_setsockopt(struct device *dev, void *obj,
 	int ret;
 
-	if (level != SOL_CAN_RAW && optname != CAN_RAW_FILTER) {
+	if (level != SOL_CAN_RAW && (optname != CAN_RAW_FILTER)) {
```
You're right, this isn't clean.

maxvankessel commented on include/net/socket_can.h:
```diff
@@ -38,11 +38,64 @@ extern "C" {
+/** error class (mask) in can_id */
+
+/** TX timeout (by netdevice driver) */
+#define CAN_ERR_TX_TIMEOUT   (0x00000001U)
```
These are aren't return codes.
These are flags set within a can_frame, as part of the identifier
`frame.can_id =  (CAN_EFF_FLAG | CAN_ERR_BUSOFF);`

This is according to the socketcan api.

maxvankessel commented on drivers/can/socket_can_generic.h:
```diff
@@ -39,40 +41,104 @@ static inline void socket_can_iface_init(struct net_if *iface)
-	ret = can_send(socket_context->can_dev,
-		       (struct zcan_frame *)pkt->frags->data,
-		       SEND_TIMEOUT, tx_irq_callback, "socket_can_send");
+	can_copy_frame_to_zframe((const struct can_frame *)pkt->frags->data,
```
Because on `zcan_frame` there can't be any error filtering done without modifying the structure.
The `can_frame` is already compatible for error filtering.

It also seems very logical that `can_frames` should come from the `socket_can` layer instead of `zcan_frames`, so I think it should be converted in the glue layer, e.g. `socket_can_generic`.

maxvankessel commented on drivers/can/socket_can_generic.h:
```diff
@@ -118,38 +191,53 @@ static struct canbus_api socket_can_api = {
-		ret = net_recv_data(socket_context->iface, pkt);
-		if (ret < 0) {
-			net_pkt_unref(pkt);
+	while (1) {
```
True, also convention states this. 
I wonder why coccinelle doesn't suggest any changes.

maxvankessel commented on subsys/net/lib/sockets/sockets_can.c:
```diff
@@ -98,19 +98,20 @@ static void zcan_received_cb(struct net_context *ctx, struct net_pkt *pkt,
-		struct zcan_frame *zframe =
-			(struct zcan_frame *)net_pkt_data(pkt);
-		struct can_frame frame;
+		struct can_frame frame = *(struct can_frame *)net_pkt_data(pkt);
```
As you can see, a `can_frame` is pushed into the message queue in socket_can_generic.h. 
We are receiving can_frame, because they can be filtered for error frames and zcan_frame can't.

maxvankessel reviewed (commented):
@jukkar thank you for your time/review.
I've placed comments on some notes to elaborate. 

On the notes where there aren't any comments, these will be fixed.

maxvankessel committed:
net: socket_can: added state change support

Complementary state propagation through sockets
Supports states: active, passive and bus-off.

Signed-off-by: Max van Kessel <max_van_kessel@msn.com>

maxvankessel committed:
net: socketcan: added can state variable to socket context

The application doesn't need to know if the state isn't changed

maxvankessel committed:
net: socketcan: changed unref statement

Unreference packet if no receiver is found or if it's cloned

alexanderwachter commented on drivers/can/socket_can_generic.h:
```diff
@@ -43,36 +44,107 @@ static inline void socket_can_iface_init(struct net_if *iface)
+	can_copy_frame_to_zframe((const struct can_frame *)pkt->frags->data,
+								&zframe);
+
+	ret = can_send(socket_context->can_dev, (struct zcan_frame *)&zframe,
```
You are changing the send call to be blocking. Is this intentional?

maxvankessel replied:
Unintentional yes,I will fix this. 
I find it a bit confusing, providing a timeout and no callback will trigger a blocking call.

alexanderwachter commented on drivers/can/socket_can_generic.h:
```diff
@@ -118,38 +197,52 @@ static struct canbus_api socket_can_api = {
 
 static struct socket_can_context socket_can_context_1;
 
+static void state_changed(enum can_state state, struct can_bus_err_cnt err_cnt)
```
Nut sure what you are doing here. Why do you want to have an Rx and Tx state? The state is the state of the controller. It doesn't mater which error counter is bigger.

maxvankessel replied:
I've created the this software according to the socketcan standard, which is implemented in the linux kernel.
According to the errors on the bus, the controller enters a warning (which is not implemented) or a passive mode.

> A node starts out in Error Active mode. When any one of the two Error Counters raises above 127, the node will enter a state known as Error Passive and when the Transmit Error Counter raises above 255, the node will enter the Bus Off state. (Source: https://www.kvaser.com/about-can/the-can-protocol/can-error-handling/)  

This is not quite complete:
>  The state is the state of the controller

It is the state according to the counters, so the counters are part of the error report. And provides the application the necessary information on the bus. A high bus load could lead to `CAN_ERR_CRTL_RX_PASSIVE`. But if your overloading your hardware transceiver , or it's (electrically) broken you might receive `CAN_ERR_CRTL_TX_PASSIVE`.

You might want to check the documentation and existing source code.

alexanderwachter replied:
The error counters do only increase on bit errors. This is according to the Bosch CAN specification [1] chapter six. The states you are describing are exactly the same as defined in the specification.
You can just translate the state to the flags used in SocketCAN

[1] http://esd.cs.ucr.edu/webres/can20.pdf

maxvankessel replied:
> When a RECEIVER detects an error, the RECEIVE ERROR COUNT will be
increased by 1, except when the detected error was a BIT ERROR during the
sending of an ACTIVE ERROR FLAG or an OVERLOAD FLAG.

This doesn't say it's only increased on bit errors.

alexanderwachter replied:
@maxvankessel Sorry, I was a little bit unclear on this. The counters only increase on errors detected, defined in chapter six. But anyway, the state you get from `enum can_state state` is the estate you should pass on.

maxvankessel replied:
It's still unclear to me what you want. 

I think the `can_state state` is passed to the `socket_can_change_state` and I do think the controller tells us there are (at least) three error possibilities `CAN_ERR_CRTL_RX_PASSIVE`, `CAN_ERR_CRTL_TX_PASSIVE` & `CAN_ERR_BUSOFF`. A generic way to find this problem is to compare error counters. Counters are not passed on.

This seems to be the way how it's done in the linux kernel.

But please, provide a suggestion.

alexanderwachter replied:
Now I got it. There are two different kinds of states defined in the Linux kernel, depending on which counter exceeded 128.

maxvankessel committed:
net: socketcan: revert on tx_irq_callback

Reverted the tx_irq_callback due to the fact its's needed to make can_send non blocking

Signed-off-by: Max van Kessel <max_van_kessel@msn.com>

maxvankessel committed:
net: socketcan: fix for incomplete filter

Removed CAN_ERR_FLAG from the filter outcome

Signed-off-by: Max van Kessel <max_van_kessel@msn.com>

maxvankessel committed:
samples: socketcan: add filter for bus error

Added different filter for other thread to demonstrate error filters

Signed-off-by: Max van Kessel <max_van_kessel@msn.com>

jukkar commented:
I did a little bit of testing with this one and it does not work as expected:
I have two `nucleo_l432kc` boards that are connected together using CAN.
What I did:
```
$ cd samples/net/sockets/can
$ cmake -B build -DBOARD=nucleo_l432kc -DCONFIG_NET_SOCKETS_CAN_RECEIVERS=2 -DCONFIG_NET_BUF_POOL_USAGE=y .
$ make -C build flash
```

Expected output from the sample looks something like this:
```
[00:00:17.400,000] <dbg> can_driver.can_stm32_send: Sending 8 bytes on CAN_1. Id: 0x1, ID type: standard, Remote Frame: no
[00:00:17.413,000] <dbg> can_driver.can_stm32_send: Using mailbox 0
[00:00:18.203,000] <dbg> can_driver.can_stm32_rx_isr_handler: Message on filter index 0
[00:00:18.213,000] <dbg> net_sock_can.zcan_received_cb: (0x20000a30): [0] ctx 0x200006d8 pkt 0x20004830 st 0
[00:00:18.226,000] <dbg> net_sock_can.zcan_received_cb: (0x20000a30): [1] ctx 0x20000760 pkt 0x20004870 st 0
[00:00:18.238,000] <inf> net_socket_can_sample: [0] CAN msg: type 0x0 RTR 0x0 EID 0x1 DLC 0x8
[00:00:18.249,000] <inf> net_socket_can_sample: Data
f0 f1 f2 f3 f4 f5 f6 f7                          |........         
[00:00:18.263,000] <inf> net_socket_can_sample: [1] CAN msg: type 0x0 RTR 0x0 EID 0x1 DLC 0x8
[00:00:18.274,000] <inf> net_socket_can_sample: Data
f0 f1 f2 f3 f4 f5 f6 f7                          |........         
```

What I am seeing with this PR looks like this:
```
[00:00:38.939,000] <dbg> can_driver.can_stm32_send: Sending 8 bytes on CAN_1. Id: 0x1, ID type: standard, Remote Frame: no
[00:00:38.952,000] <dbg> can_driver.can_stm32_send: Using mailbox 0
[00:00:39.128,000] <dbg> can_driver.can_stm32_rx_isr_handler: Message on filter index 0
[00:00:39.139,000] <dbg> net_sock_can.zcan_received_cb: (0x200009d0): [0] ctx 0x20000678 pkt 0x20004b68 st 0
[00:00:39.151,000] <inf> net_socket_can_sample: [0] CAN msg: type 0x0 RTR 0x0 EID 0x1 DLC 0x8
[00:00:39.162,000] <inf> net_socket_can_sample: Data
f0 f1 f2 f3 f4 f5 f6 f7                          |........         
```
So the socket 1 output is missing.

maxvankessel commented:
@jukkar I believe thats correct, each error filter also uses a `receiver`. So on each thread there are 2 filters. Which makes the total of 4 filters if you use 2 threads. 

So `CONFIG_NET_SOCKETS_CAN_RECEIVERS=4` will make this sample more as you expect

jukkar commented:
> So `CONFIG_NET_SOCKETS_CAN_RECEIVERS=4` will make this sample more as you expect

Unfortunately I did not see any change in behavior when setting the receiver count to 4:
```
[00:00:14.344,000] <dbg> can_driver.can_stm32_send: Sending 8 bytes on CAN_1. Id: 0x1, ID type: standard, Remote Frame: no
[00:00:14.358,000] <dbg> can_driver.can_stm32_send: Using mailbox 0
[00:00:14.988,000] <dbg> can_driver.can_stm32_rx_isr_handler: Message on filter index 0
[00:00:14.999,000] <dbg> net_sock_can.zcan_received_cb: (0x200009ec): [0] ctx 0x20000694 pkt 0x200043a8 st 0
[00:00:15.012,000] <inf> net_socket_can_sample: [0] CAN msg: type 0x0 RTR 0x0 EID 0x1 DLC 0x8
[00:00:15.024,000] <inf> net_socket_can_sample: Data
f0 f1 f2 f3 f4 f5 f6 f7                          |........         
```

jukkar reviewed (changes requested):
There is something wrong when running this PR with socketcan sample when receiver count > 1, that issue needs to be solved before this can be merged.

maxvankessel commented:
@jukkar Im not sure why you are seeing this, but this is my output
```
[00:00:02.006,000] <dbg> net_sock_can.can_register_filters: (0x20000d8c): Registering 1 filters
[00:00:02.017,000] <dbg> net_sock_can.can_register_receiver: (0x20000d8c): Max 4 receivers
[00:00:02.028,000] <dbg> can_driver.can_stm32_set_filter: Setting filter ID: 0x1, mask: 0x7ff
[00:00:02.039,000] <dbg> can_driver.can_stm32_set_filter: Filter type: standard ID without mask (1)
[00:00:02.050,000] <dbg> can_driver.can_stm32_set_filter: Filter set! Filter number: 0 (index 0)
[00:00:02.062,000] <dbg> net_sock_can.can_register_receiver: (0x20000d8c): Max 4 receivers
[00:00:02.073,000] <dbg> net_socket_can_sample.setup_socket: Started socket CAN TX thread
[00:00:02.083,000] <inf> net_socket_can_sample: 1st RX fd 0
[00:00:02.092,000] <dbg> net_sock_can.can_register_filters: (0x20000d8c): Registering 1 filters
[00:00:02.103,000] <dbg> net_sock_can.can_register_receiver: (0x20000d8c): Max 4 receivers
[00:00:02.114,000] <dbg> net_sock_can.can_register_receiver: (0x20000d8c): Max 4 receivers
[00:00:02.124,000] <inf> net_socket_can_sample: 2nd RX fd 1
[00:00:02.133,000] <dbg> net_socket_can_sample.rx: [0] Waiting CAN data...
[00:00:02.142,000] <dbg> net_socket_can_sample.rx: [1] Waiting CAN data...
[00:00:02.346,000] <dbg> can_driver.can_stm32_rx_isr_handler: Message on filter index 0
[00:00:02.356,000] <dbg> net_sock_can.zcan_received_cb: (0x20000aac): [0] ctx 0x20000754 pkt 0x20005058 st 0
[00:00:02.369,000] <dbg> net_sock_can.zcan_received_cb: (0x20000aac): [2] ctx 0x200007dc pkt 0x20005018 st 0
[00:00:02.381,000] <inf> net_socket_can_sample: [0] CAN msg: type 0x0 RTR 0x0 EID 0x1 DLC 0x5
[00:00:02.392,000] <inf> net_socket_can_sample: Data
11 22 33 44 55                                   |."3DU            
[00:00:02.406,000] <inf> net_socket_can_sample: [1] CAN msg: type 0x0 RTR 0x0 EID 0x1 DLC 0x5
[00:00:02.417,000] <inf> net_socket_can_sample: Data
11 22 33 44 55                                   |."3DU            
[00:00:03.073,000] <dbg> net_socket_can_sample.tx: Sending CAN data...

```
I'm using a CAN transceiver (tja1050) with with a sysWORKXX canusb device. 
With the following command: `cangen can0 -g 1000 -I 1 -L i -D 1122334455667788`

I've set `CONFIG_NET_SOCKETS_CAN_RECEIVERS=4` 

Could you supply some more information?

jukkar commented:
> I'm using a CAN transceiver (tja1050) with with a sysWORKXX canusb device.

I was using two nucleo_l432kc boards connected together and running the `samples/net/sockets/can` sample app with CONFIG_NET_SOCKETS_CAN_RECEIVERS=2 config option. With that sample and board, I am seeing a regression when I tested using this PR.

maxvankessel commented:
Without any transceivers? Just rx -> tx and vice versa?

jukkar commented:
> Without any transceivers? Just rx -> tx and vice versa?

No. Each board has this transceiver http://skpang.co.uk/catalog/canbus-can-fd-breakout-board-5v-p-242.html

maxvankessel commented:
So @jukkar it's been a while, I had to find some compatible hardware to communicate with.
I couldn't get any CAN sample to work with my L476RG, but I have found an F091CC. 
Which as suspected, works perfectly:
```
[00:07:47.691,000] <dbg> can_driver.can_stm32_rx_isr_handler: Message on filter index 0
[00:07:47.692,000] <inf> net_socket_can_sample: [0] CAN msg: type 0x0 RTR 0x0 EID 0x1 DLC 0x8
[00:07:47.692,000] <inf> net_socket_can_sample: Data
f0 f1 f2 f3 f4 f5 f6 f7                          |........         
[00:07:47.692,000] <inf> net_socket_can_sample: [1] CAN msg: type 0x0 RTR 0x0 EID 0x1 DLC 0x8
[00:07:47.692,000] <inf> net_socket_can_sample: Data
f0 f1 f2 f3 f4 f5 f6 f7                          |........      
```

Could you please test again, and provide more information on your problem. 
Or maybe someone else could test this PR?

jukkar commented:
Tried this again with latest upstream. I did not see any progress
```
$ cd samples/net/sockets/can
$ cmake -DCONFIG_NET_SOCKETS_CAN_RECEIVERS=2 -DBOARD=nucleo_l432kc -B build .
$ make -C build flash

[00:01:55.746,000] <dbg> can_driver.can_stm32_send: Sending 8 bytes on CAN_1. Id: 0x1, ID type: standard, Remote Frame: no
[00:01:55.760,000] <dbg> can_driver.can_stm32_send: Using mailbox 0
[00:01:56.521,000] <dbg> can_driver.can_stm32_rx_isr_handler: Message on filter index 0
[00:01:56.532,000] <dbg> net_sock_can.zcan_received_cb: (0x200009d0): [0] ctx 0x20000678 pkt 0x20004b68 st 0
[00:01:56.545,000] <inf> net_socket_can_sample: [0] CAN msg: type 0x0 RTR 0x0 EID 0x1 DLC 0x8
[00:01:56.556,000] <inf> net_socket_can_sample: Data
f0 f1 f2 f3 f4 f5 f6 f7                          |........         
[00:01:56.769,000] <dbg> can_driver.can_stm32_send: Sending 8 bytes on CAN_1. Id: 0x1, ID type: standard, Remote Frame: no
[00:01:56.783,000] <dbg> can_driver.can_stm32_send: Using mailbox 0
[00:01:57.542,000] <dbg> can_driver.can_stm32_rx_isr_handler: Message on filter index 0
[00:01:57.553,000] <dbg> net_sock_can.zcan_received_cb: (0x200009d0): [0] ctx 0x20000678 pkt 0x20004b68 st 0
[00:01:57.566,000] <inf> net_socket_can_sample: [0] CAN msg: type 0x0 RTR 0x0 EID 0x1 DLC 0x8
[00:01:57.577,000] <inf> net_socket_can_sample: Data
f0 f1 f2 f3 f4 f5 f6 f7                          |........         
```
Only first socket is in use and the other does not get any data. Any idea what is going on here?

maxvankessel commented:
It should be build with at least 4 receivers, for complete sample
'-DCONFIG_NET_SOCKETS_CAN_RECEIVERS=2'

jukkar commented:
> It should be build with at least 4 receivers, for complete sample:
> `-DCONFIG_NET_SOCKETS_CAN_RECEIVERS=4`
> 
> Make sure to clean the project thoroughly.

I do not follow your logic here. If there are only two socket CAN listeners in the application (like currently in socket CAN sample), why would user need to define four listeners?

maxvankessel commented:
Sorry to say @jukkar but this is your logic, the code is written by you. 

It is due to the fact that only one receiver can hold one filter.
`can_register_filters` which calls `can_register_receiver` which searches for the next empty spot in the receiver list (`receivers[i].ctx != NULL` then `continue`).

So the samples creates 2 filters per socket, and there are 2 sockets, 1 for the main_thread and 1 for the spawned thread. So a total of 4 filters, so a total of 4 receivers.

I modified the sample to support error (one) filter on the spawned thread and (one) on the main thread, but only if there are more than 2 receivers, if you specify 3 receivers it would create 2 filters on the spawned thread and only 1 on the main thread.

alexanderwachter commented:
> It should be build with at least 4 receivers, for complete sample:
> `-DCONFIG_NET_SOCKETS_CAN_RECEIVERS=4`
> 
> Make sure to clean the project thoroughly.

If we have four receivers now (two for frames and two for error frames), you should add it to the samples defconfig, so the samples work out of the box.

jukkar commented:
I tried this once more. Compiled the socketcan sample like this `cmake -Bbuild . -DBOARD=nucleo_l432kc -DCONFIG_NET_SOCKETS_CAN_RECEIVERS=10 -DCONFIG_NET_SHELL=y` and now I am seeing expected logging
```
[00:01:44.870,000] <inf> net_socket_can_sample: [0] CAN msg: type 0x0 RTR 0x0 EID 0x1 DLC 0x8
[00:01:44.882,000] <inf> net_socket_can_sample: Data
f0 f1 f2 f3 f4 f5 f6 f7                          |........         
[00:01:44.896,000] <inf> net_socket_can_sample: [1] CAN msg: type 0x0 RTR 0x0 EID 0x1 DLC 0x8
[00:01:44.908,000] <inf> net_socket_can_sample: Data
f0 f1 f2 f3 f4 f5 f6 f7                          |........         
```
The problem with this PR is that the sample application is interested only for two socketcan sockets, thus the expectation is that it would work just fine with `CONFIG_NET_SOCKETS_CAN_RECEIVERS=2`, but it does not atm, one needs to increase the number of receivers much higher.

maxvankessel commented:
> > It should be build with at least 4 receivers, for complete sample:
> > `-DCONFIG_NET_SOCKETS_CAN_RECEIVERS=4`
> > Make sure to clean the project thoroughly.
> 
> If we have four receivers now (two for frames and two for error frames), you should add it to the samples defconfig, so the samples work out of the box.

Oke, done
